src/uct/tcp/tcp_cm.c

```diff
@@ -8,100 +8,299 @@
+{
+    uct_tcp_iface_t *iface         = ucs_derived_of(ep->super.super.iface,
+                                                    uct_tcp_iface_t);
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_REQ
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    size_t addr_size;
+    ucs_status_t status;
+
+    status = ucs_sockaddr_sizeof((const struct sockaddr*)&iface->config.ifaddr,
+                                 &addr_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    memcpy(&conn_pkt.iface_addr, &iface->config.ifaddr, addr_size);
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection request sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_recv_conn_req(uct_tcp_ep_t *ep,
+                                             struct sockaddr *peer_addr)
+{
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_assertv(conn_pkt.event == UCT_TCP_CM_CONN_REQ, "ep=%p", ep);
+    *peer_addr = conn_pkt.iface_addr;
+
+    ucs_debug("tcp_ep %p: received the connection request from %s",
+              ep, ucs_sockaddr_str(peer_addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_ACK
+    };
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (ucs_unlikely(status != UCS_OK)) {
+        return status;
+    }
+
+    ucs_debug("tcp_ep %p: conection ack sent to %s",
+              ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                   UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cn_recv_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+
+    ucs_assertv((status != UCS_OK) ||
+                (conn_pkt.event == UCT_TCP_CM_CONN_ACK), "ep=%p", ep);
+    return status;
+}
+
+unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
+{
+    unsigned ret = 0;
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
     ucs_status_t status;
 
     status = ucs_socket_connect_nb_get_status(ep->fd);
-    if (status == UCS_INPROGRESS) {
+    if (ucs_likely(status == UCS_OK)) {
+        ret++;
```

Comment:
need to check whether it is needed, because here you do not send/receive anything. IMO, we do not need to increase progress counter here.

Reply:
agree, this is an internal event

Reply:
the comment was not applied

Reply:
sorry, the last commit was added with forced-push by mistake. Please, take a look through "Files changed"

Reply:
IMO, if internal event progresses some protocol then better to increment the counter to notify user that ucp_progress have done some job, the same as, poll_cq > 0 may don't complete any user's request, e.g. RNDV or fragmented protocols

Reply:
but `ucs_socket_connect_nb_get_status` is not the case